test/dynamo/test_misc.py

```diff
@@ -311,6 +311,17 @@ def fn(x):
         r2 = opt_fn(i)
         self.assertTrue(same(r1, r2))
 
+    def test_tensor_iter(self):
+        def fn(x):
+            for y in x:
+                y.add_(1.0)
+            return y
+
+        # expect extra size node for dynamic
+        torch._dynamo.testing.standard_test(
+            self, fn, 1, expected_ops=20, expected_ops_dynamic=21
+        )
```

Comment:
can you add a guard failure test? `evaluate_expr` below incurs a guard, which is what we want, but I would like to see something like:

1) trace fn with X
2) run fn with Y where Y.size != X.size
3) run fn with Z where Z.size != Y.size != X.size

A) the guard failure fn should report that a guard failed because the value from Z is not the value from Y
B) we should also add comparisons to real values between 2 and 3 above to ensure we are not baking in /compiling in an unwound loop of a fixed size